sdk/core/azure-core/inc/azure/core/http/websockets/websockets_transport.hpp

```diff
@@ -0,0 +1,252 @@
+      /**
+       * @brief Indicates that the frame is either the complete binary frame to be sent
+       * to the remote node or the final frame of a multipart message.
+       */
+      FrameTypeBinary,
+      /**
+       * @brief Indicates that the frame is a partial binary frame - it is NOT the
+       * complete frame to be sent to the remote node.
+       */
+      FrameTypeBinaryFragment,
+
+      FrameTypeClosed,
+    };
+    /**
+     * @brief Destructs `%WebSocketTransport`.
+     *
+     */
+    virtual ~WebSocketTransport() {}
+
+    /**
+     * @brief Determines if the transport natively supports WebSockets or not.
+     *
+     * @returns true if the transport has native websocket support, false otherwise.
+     */
+    virtual bool HasNativeWebsocketSupport() = 0;
+
+    /**************/
+    /* Native WebSocket support functions*/
+    /**************/
+    /**
+     * @brief Gracefully closes the WebSocket, notifying the remote node of the close reason.
+     *
+     * @param status Status value to be sent to the remote node. Application defined.
+     * @param disconnectReason UTF-8 encoded reason for the disconnection. Optional.
+     * @param context Context for the operation.
+     */
+    virtual void CloseSocket(
+        uint16_t status,
+        std::string const& disconnectReason,
+        Azure::Core::Context const& context)
+        = 0;
+
+    /**
+     * @brief Closes the WebSocket.
+     *
+     * Does not notify the remote endpoint that the socket is being closed.
+     *
+     */
+    virtual void Close() = 0;
+
+    /**
+     * @brief Retrieve the information associated with a WebSocket close response.
+     *
+     * @param context Context for the operation.
+     *
+     * @returns a tuple containing the status code and string.
+     */
+    virtual std::pair<uint16_t, std::string> GetCloseSocketInformation(
+        Azure::Core::Context const& context)
+        = 0;
+
+    /**
+     * @brief Send a frame of data to the remote node.
+     *
+     * @param frameType Frame type sent to the server, Text or Binary.
+     * @param frameData Frame data to be sent to the server.
+     * @param context Context for the operation.
+     */
+    virtual void SendFrame(
+        WebSocketFrameType frameType,
+        std::vector<uint8_t> const& frameData,
+        Azure::Core::Context const& context)
+        = 0;
+
+    /**
+     * @brief Receive a frame from the remote WebSocket server.
+     *
+     * @param context Context for the operation.
+     *
+     * @returns a tuple containing the Frame data received from the remote server and the type of
+     * data returned from the remote endpoint
+     */
+    virtual std::pair<WebSocketFrameType, std::vector<uint8_t>> ReceiveFrame(
+        Azure::Core::Context const& context)
+        = 0;
+
+    /**************/
+    /* Non Native WebSocket support functions */
+    /**************/
+
+    // Implement a buffered stream reader
+    class BufferedStreamReader {
```

Comment:
An `XxxTransport` has the job of sending and receiving messages over a TCP socket.

For an `HttpTransport`, the only meaningful verb is "Send" since Http is a simple request/response protocol (it's not, but the `HttpTransport` abstracts away that implementation detail).
 
The `WebSocketTransport` is no more stateful than the `HttpTransport` is, but it supports more verbs - `SendBuffer` and `ReadFromSocket`. 

The transport is a shared pointer because there are multiple objects within the `WebSocket` which access the transport. And storing multiple references to the single unique_ptr is a *horrible* design.

Reply:
> The WebSocketTransport is no more stateful than the HttpTransport is

Let me tell you why I see this is not correct.  I can see that a `BufferedStreamReader` uses 2 things, an in-memory buffer and a `WebSocketTransport`.  Whenever there is no more data in the in-memory buffer, it uses the `WebSocketTransport` to read more bytes from the network.  And that's the right point where it is different than the `HttpTransport`.
The `HttpTransport` can open a connection and send bytes, but it is not used to keep reading bytes from the network.  Instead, the `send()` function returns a `rawResponse` which contains a `BodyStream` that we use to read form the network. That body streams works on top of a network connection (curl or winhttp), but that connection is not the `HttpTransport`.

Basically, the HttpTransport is used to create a connection and/or a network session. In order to read bytes from network, we don't do `HttpTransport.ReadBytes()`

Here's how a WebsocketTransport would be similar to HttpTransport:

```cpp
WebSocketTransport wsTransport;

auto wsSession = wsTransport.Open(url, options, etc);
wsSession.ReadBytes(...)
wsSession.SendFrame(...)
```

This way, we can have only one instance of `wsTransport` within an application that is used to produce a `wsSession` everytime.  And you can call `wsTransport.Open(..)` in a thread-safe manner.  

With the current implementation, the program needs to have one instance of `WebSocketTransport` for every connection.  And that instance has `.Open(..)` and `.Send(..)`.  One function needs to check if the other has been called before (check state).

Reply:
There is no "BufferedStreamReader" class, so I'm not quite sure what you're referring to.

There was an unreferenced "BufferedStreamReader" declaration that was checked in accidentally, but I removed that declaration.

Reply:
Ping :).